File: Daemon/Watchdog.fs
namespace Daemon

open System.Timers
open System
open System.Net

type Watchdog(ms, server:HttpServer) =

    let timer = new Timer(ms)

    let OnTimerEvent args = 
        Console.WriteLine "Watchdog is stopping the server"
        server.Stop

    do
        timer.AutoReset <- false
        timer.Elapsed.Add(OnTimerEvent)
        timer.Start()

    member this.StayAlive (context:HttpListenerContext) = 
        timer.Stop()
        timer.Start()
        Http.SendPlainText context.Response "OK"

File: Daemon/Watchdog.cs
using Daemon.Utils;
using System;
using System.Net;
using System.Timers;

namespace Daemon
{
    class Watchdog
    {
        public Watchdog(int ms)
        {
            timer = new Timer(ms);
            timer.AutoReset = false;
            timer.Elapsed += OnTimerEvent;
        }

        public void BindTo(HttpServer server)
        {
            this.server = server;

            server.InstallHandler("/stayAlive", StayAlive);

            timer.Start();
        }

        private void OnTimerEvent(object source, ElapsedEventArgs e)
        {
            Console.WriteLine("Server stopped by watchdog");
            server.Stop();
        }

        private void StayAlive(HttpListenerContext context)
        {
            timer.Stop();
            timer.Start();
            context.Response.SendPlainText("OK");
        }

        private Timer timer;
        private HttpServer server;
    }
}
